stage1-fly.aci panics with empty exec

I've got an ACI without `exec` specified in the manifest.  I can `rkt run` it if I provide `--exec=/bin/bash`.  If I leave off the `--exec`, stage1-coreos.aci will complain:

```
# rkt run --interactive example.com/aci/centos7,version=765cc162-18 
image: using image from local store for image name coreos.com/rkt/stage1-coreos:1.0.0
image: using image from local store for image name example.com/aci/centos7,version=765cc162-18
networking: loading networks from /etc/rkt/net.d
networking: loading network default with type ptp
stage1: failed to configure systemd: image "example.com/aci/centos7" has an empty "exec" (try --exec=BINARY)
```

If I use stage1-fly.aci, I get a panic:

```
# rkt run --stage1-from-dir=stage1-fly.aci  --interactive example.com/aci/centos7,version=765cc162-18 
image: using image from file /usr/lib/rkt/stage1-images/stage1-fly.aci
image: using image from local store for image name example.com/aci/centos7,version=765cc162-18
panic: runtime error: index out of range

goroutine 1 [running]:
main.stage1.func1(0x0, 0x0)
	/home/iaguis/work/coreos/go/src/github.com/coreos/rkt/release-build/gopath/src/github.com/coreos/rkt/stage1_fly/run/main.go:334 +0xba
github.com/coreos/rkt/stage1/common.WithClearedCloExec(0x5, 0xc8200d7c28, 0x0, 0x0)
	/home/iaguis/work/coreos/go/src/github.com/coreos/rkt/release-build/gopath/src/github.com/coreos/rkt/stage1/common/run.go:37 +0x9b
main.stage1(0xc8200a4030)
	/home/iaguis/work/coreos/go/src/github.com/coreos/rkt/release-build/gopath/src/github.com/coreos/rkt/stage1_fly/run/main.go:335 +0x1dd3
main.main()
	/home/iaguis/work/coreos/go/src/github.com/coreos/rkt/release-build/gopath/src/github.com/coreos/rkt/stage1_fly/run/main.go:353 +0xaa

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/lib/go/src/runtime/asm_amd64.s:1721 +0x1
```

--- a/stage1/common/types/pod.go
+++ b/stage1/common/types/pod.go
@@ -37,6 +37,18 @@ type Pod struct {
 	Networks           []string
 }
 
+// AppNameToImageName takes the name of an app in the Pod and returns the name
+// of the app's image. The mapping between these two is populated when a Pod is
+// loaded (using LoadPod).
+func (p *Pod) AppNameToImageName(appName types.ACName) types.ACIdentifier {
+	image, ok := p.Images[appName.String()]
+	if !ok {
+		// This should be impossible as we have updated the map in LoadPod().
+		panic(fmt.Sprintf("No images for app %q", appName.String()))
+	}
+	return image.Name
+}
+
 // LoadPod loads a Pod Manifest (as prepared by stage0) and
 // its associated Application Manifests, under $root/stage1/opt/stage1/$apphash
 func LoadPod(root string, uuid *types.UUID) (*Pod, error) {
@@ -58,24 +70,24 @@ func LoadPod(root string, uuid *types.UUID) (*Pod, error) {
 	p.Manifest = pm
 
 	for i, app := range p.Manifest.Apps {
-		ampath := common.ImageManifestPath(p.Root, app.Name)
-		buf, err := ioutil.ReadFile(ampath)
+		impath := common.ImageManifestPath(p.Root, app.Name)
+		buf, err := ioutil.ReadFile(impath)
 		if err != nil {
-			return nil, errwrap.Wrap(fmt.Errorf("failed reading app manifest %q", ampath), err)
+			return nil, errwrap.Wrap(fmt.Errorf("failed reading image manifest %q", impath), err)
 		}
 
-		am := &schema.ImageManifest{}
-		if err = json.Unmarshal(buf, am); err != nil {
-			return nil, errwrap.Wrap(fmt.Errorf("failed unmarshalling app manifest %q", ampath), err)
+		im := &schema.ImageManifest{}
+		if err = json.Unmarshal(buf, im); err != nil {
+			return nil, errwrap.Wrap(fmt.Errorf("failed unmarshalling image manifest %q", impath), err)
 		}
 
 		if _, ok := p.Images[app.Name.String()]; ok {
 			return nil, fmt.Errorf("got multiple definitions for app: %v", app.Name)
 		}
 		if app.App == nil {
-			p.Manifest.Apps[i].App = am.App
+			p.Manifest.Apps[i].App = im.App
 		}
-		p.Images[app.Name.String()] = am
+		p.Images[app.Name.String()] = im
 	}
 
 	return p, nil

--- a/stage1/init/common/pod.go
+++ b/stage1/init/common/pod.go
@@ -234,12 +234,7 @@ func generateGidArg(gid int, supplGid []int) string {
 func appToSystemd(p *stage1commontypes.Pod, ra *schema.RuntimeApp, interactive bool, flavor string, privateUsers string) error {
 	app := ra.App
 	appName := ra.Name
-	image, ok := p.Images[appName.String()]
-	if !ok {
-		// This is impossible as we have updated the map in LoadPod().
-		panic(fmt.Sprintf("No images for app %q", ra.Name.String()))
-	}
-	imgName := image.Name
+	imgName := p.AppNameToImageName(appName)
 
 	if len(app.Exec) == 0 {
 		return fmt.Errorf(`image %q has an empty "exec" (try --exec=BINARY)`, imgName)

--- a/stage1_fly/run/main.go
+++ b/stage1_fly/run/main.go
@@ -206,8 +206,18 @@ func stage1() int {
 		return 1
 	}
 
+	// Sanity checks
 	if len(p.Manifest.Apps) != 1 {
 		log.Printf("flavor %q only supports 1 application per Pod for now", flavor)
+		return 1
+	}
+
+	ra := p.Manifest.Apps[0]
+
+	imgName := p.AppNameToImageName(ra.Name)
+	args := ra.App.Exec
+	if len(args) == 0 {
+		log.Printf(`image %q has an empty "exec" (try --exec=BINARY)`, imgName)
 		return 1
 	}
 
@@ -224,14 +234,13 @@ func stage1() int {
 	}
 
 	env := []string{"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"}
-	for _, e := range p.Manifest.Apps[0].App.Environment {
+	for _, e := range ra.App.Environment {
 		env = append(env, e.Name+"="+e.Value)
 	}
 
-	args := p.Manifest.Apps[0].App.Exec
-	rfs := filepath.Join(common.AppPath(p.Root, p.Manifest.Apps[0].Name), "rootfs")
+	rfs := filepath.Join(common.AppPath(p.Root, ra.Name), "rootfs")
 
-	argFlyMounts, err := evaluateMounts(rfs, string(p.Manifest.Apps[0].Name), p)
+	argFlyMounts, err := evaluateMounts(rfs, string(ra.Name), p)
 	if err != nil {
 		log.PrintE("can't evaluate mounts", err)
 		return 1

